### ISSUE TITLE
Super slow repo lookup for repo paths on sourcegraph.com

### ISSUE THREAD

**Issue by rvantonder** [sourcegraph/sourcegraph#12889]
TL;DR: why is searching for 
1. [`repo:github.com/sourcegraph/sourcegraph`](https://sourcegraph.com/search?q=repo:github.com/sourcegraph/sourcegraph&patternType=literal) fast (<200ms) and 
1. [`repo:github\.com/sourcegraph/sorucegraph`](https://sourcegraph.com/search?q=repo:github%5C.com/sourcegraph/sourcegraph&patternType=literal) dirt slow (> 5s) ? 

Or, alternatively, why:

1. [`repo:github.com/sourcegraph/sourcegraph(.*)`](https://sourcegraph.com/search?q=repo:github.com/sourcegraph/sourcegraph%28.*%29&patternType=literal) fast (<200ms) and 
1. [`repo:github\.com/sourcegraph/sourcegraph(.*)`](https://sourcegraph.com/search?q=repo:github%5C.com/sourcegraph/sourcegraph%28.*%29&patternType=literal) dirt slow ( > 4s)?

Why does using `\.` instead of `.` cause this big discrepancy? ~My guess is that it's again some index/DB slowness that is possibly using `LIKE` instead of regex when the whole string is a literal. Maybe @keegancsmith investigation about `lower` is related.~

**Edit: Answer: we were applying a heuristic that appended `^` on sourcegraph.com only if the repo string contains exactly `github.com`, (and not `github\.com`).**

---

Outdated:

**Note:** prepending a `^` to both of the dirt slow queries make them fast for some nebulous reason.

**Note:** I'm surprised the above case isn't solved with the [repo name index](https://github.com/sourcegraph/sourcegraph/commit/5f278c6ce9901f460a766cef7fa9d57aabbad10a) but turns out it doesn't. The repo index _does_ fix queries like [`r:^github.com/sourcegraph/`](https://sourcegraph.com/search?q=r:%5Egithub.com/sourcegraph/&patternType=literal) which previously took > 5 seconds and now take 100ms. Note that this query is still regex. The problem above stems from a string that has no (visible) regex, however, I do think the above is generally expanded to `(.*)literal(.*)` or `(.*?)literal(.*?)`.

**Note:** This issue affects the globbing in a bad way because we effectively search for `github\.com/sourcegraph/sourcegraph` (i.e., the literal) and not `github.com/sourcegraph/sourcegraph`. That's how I noticed this issue.

CC @sourcegraph/search

**Comment by rvantonder** [comment]
Ah OK, I got it. So, we _do_ do the optimization prepending the `^` if we see the string `github.com`, which is why those queries, and the ones that do _not_ contain an escaped `.`:

https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/search.go#L999-1001.

But when the query is `github\.com` (either entered manually, or, as we convert it in globbing) this "optimization" isn't done, so all queries turn out slow. I will fix it so that the escaped variety also prepends a `^`.

**Comment by keegancsmith** [comment]
Yeah I deep dived into this last week. I see you solved this particular issue, but some general info. My theory for why postgres is slow from the sourcecode I have read so far: trigram selection on your query is unaware of the size of the trigram's posting list. So you will see lots of weird performance behaviours. For example the regex `sourcegraph` is 5x faster to search for than `github\.com/sourcegraph`, even though the latter is more specific. This is because we load in posting lists for all the trigrams in `github.com` which is essentially covering the whole table, then need to intersect them with the smaller postings lists for `/sourcegraph`. The former just needs to load in the smaller posting lists for `/sourcegraph`. I wouldn't be surprised at `github.com/sourcegraph` being faster than `github\.com/sourcegraph` since we won't load the posting lists for trigrams containing the any operator (`.`).

The reason `^` works so well is we get to use the btree to do prefix lookups.

**Comment by rvantonder** [comment]
It's cool you're looking into the deeper parts of "can we make this faster in the case where we don't use the `^` for prefixes?". The trigram issue sounds similar to the thing I ran into before that sent me off on that repo rabbit hole.

But yeah for this particular issue I was just very confused about the difference for the example queries I put above, something that should be the same. The original `^github.com` trick has been there for a couple of years, so when I discovered that, it was easy to just accept that we need to do this right now.
